File: vPython/backend/app/models/gasto.py
"""
Modelo de Gasto
"""
from sqlalchemy import Column, Integer, String, Boolean, Numeric, Date, DateTime
from sqlalchemy.sql import func
from app.db.base import Base


class Gasto(Base):
    """Modelo de gastos - Equivalente a la tabla gastos"""
    __tablename__ = "gastos"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    concepto = Column(String, nullable=False)
    monto = Column(Numeric(10, 2), nullable=False)
    tipo_gasto = Column(String, nullable=False)
    forma_pago = Column(String, nullable=False)
    mes = Column(String, nullable=False)
    anio = Column(Integer, nullable=False)
    fecha_cargo = Column(Date, nullable=False)
    fecha_pago = Column(Date, nullable=False)
    categoria = Column(String, nullable=False)
    a_pagos = Column(Boolean, default=False)
    no_mens = Column(Integer, default=0)
    total_meses = Column(Integer, default=0)
    tag = Column(String, default="NA")
    se_divide = Column(Boolean, default=False)
    gasto_x_mes = Column(String, default="NA")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def __repr__(self):
        return f"<Gasto(id={self.id}, concepto='{self.concepto}', monto={self.monto})>"

File: vTS/shared/types/gasto.ts
// Tipos principales basados en la hoja "unificado_v4" del Excel

export interface Gasto {
  id: number;
  concepto: string;
  monto: number;
  tipo_gasto: TipoGasto;
  forma_pago: string;
  mes: string;
  anio: number;
  fecha_cargo: Date;
  fecha_pago: Date;
  categoria: Categoria;
  a_pagos: boolean;
  no_mens: number;
  total_meses: number;
  tag: string;
  se_divide: boolean;
  gasto_x_mes: string;
  created_at?: Date;
  updated_at?: Date;
}

export type TipoGasto = 'MSI' | 'Variable' | 'Fijo';

export type Categoria = 'E' | 'I'; // E = Egreso, I = Ingreso

export interface GastoCreateInput {
  concepto: string;
  monto: number;
  tipo_gasto: TipoGasto;
  forma_pago: string;
  mes: string;
  anio: number;
  fecha_cargo: Date;
  fecha_pago: Date;
  categoria: Categoria;
  a_pagos: boolean;
  no_mens: number;
  total_meses: number;
  tag: string;
  se_divide: boolean;
  gasto_x_mes: string;
}

export interface GastoUpdateInput extends Partial<GastoCreateInput> {
  id: number;
}

export interface GastoFilters {
  tipo_gasto?: TipoGasto[];
  categoria?: Categoria[];
  forma_pago?: string[];
  mes?: string[];
  año?: number[];
  fecha_desde?: Date;
  fecha_hasta?: Date;
  a_pagos?: boolean;
  se_divide?: boolean;
  tag?: string[];
}

// Tipos para las tablas filtradas
export interface GastoMSIMCI extends Pick<Gasto, 
  'id' | 'concepto' | 'monto' | 'forma_pago' | 'fecha_cargo' | 
  'total_meses' | 'no_mens' | 'gasto_x_mes' | 'tipo_gasto'
> {}

// Tipos para el dashboard
export interface DashboardData {
  total_gastos: number;
  total_ingresos: number;
  balance_mensual: number;
  gastos_por_tipo: Record<TipoGasto, number>;
  gastos_por_categoria: Record<Categoria, number>;
  gastos_por_mes: Array<{
    mes: string;
    anio: number;
    total: number;
  }>;
  deuda_total: number;
  diferencia_lulu: number;
}

// Tipos para balance
export interface BalanceItem {
  id: number;
  tipo: 'D' | 'C'; // D = Debe, C = Crédito
  concepto: string;
  monto: number;
  deben_ser: number;
  created_at?: Date;
  updated_at?: Date;
}

// Tipos para deudas
export interface DeudaItem {
  id: number;
  tipo: 'T' | 'O'; // T = Tarjeta, O = Otro
  item: string;
  monto: number;
  fecha: Date;
  created_at?: Date;
  updated_at?: Date;
}

// Tipos para listas/catálogos
export interface TipoGastoItem {
  id: string;
  nombre: TipoGasto;
  descripcion?: string;
}

export interface FormaPagoItem {
  id: string;
  nombre: string;
  descripcion?: string;
}

export interface CategoriaItem {
  id: string;
  codigo: Categoria;
  descripcion: string;
}
